### ISSUE TITLE
Performance of the Per site switches counters

### ISSUE THREAD

**Issue by gwarser** [uBlockOrigin/uBlock-issues#756]
<details><summary>Prerequisites
</summary>

### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [x] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

</details>

### Description

High cpu usage, over 4s in `querySelectorAll`, every time uBO popup is opened when on https://github.com/gorhill/uBlock/commit/3224d9b5cc72a3db007f4c681bcd622ecdbd923f

Even on https://github.com/orgs/uBlockOrigin/teams/ublock-issues-volunteers I see times over 1s.

Example profile:
https://profiler.firefox.com/public/540cf7624a701b19a0927eef13c3fc5283aa364e/calltree/?globalTrackOrder=0-1&implementation=js&localTrackOrderByPid=17028-0~17307-0~&thread=1&v=4

Nothing seems to be freezing or locking, but it's just waste of resources. Hmmm... I think I had problems on my phone - opening popup can be slow/jaggy, but I will need to profile this specifically.

Even Chrome is not very good, over 1s: 
![image](https://user-images.githubusercontent.com/886325/66702561-4d5deb00-ed09-11e9-8b2e-c49d95a21c54.png)

Internal discussion: https://github.com/orgs/uBlockOrigin/teams/ublock-issues-volunteers/discussions/146

### A specific URL where the issue occurs

https://github.com/gorhill/uBlock/commit/3224d9b5cc72a3db007f4c681bcd622ecdbd923f

### Steps to Reproduce

1. Open page with large DOM, example https://github.com/gorhill/uBlock/commit/3224d9b5cc72a3db007f4c681bcd622ecdbd923f
2. Click uBO button to open popup panel

### Expected behavior:

No impact on CPU usage

### Actual behavior:

High CPU usage for up to few secons

### Your environment

* uBlock Origin version: 1.22.5rc3
* Browser Name and version: Nightly 71, Chrome 77
* Operating System and version: Manjaro

**Comment by gorhill** [comment]
> checkmark overlay once, when something at all has been blocked

Just finding out if one single element has been filtered potentially requires testing all injected cosmetic filters -- which is what happens in the mentioned test case -- so just finding out if only one element is filtered is no better than counting them all when no elements are filtered.

**Comment by gwarser** [comment]
Then only [this case](https://profiler.firefox.com/public/5d43f04f297effff8d644f853346b7f452bac9d0/calltree/?globalTrackOrder=0&implementation=js&invertCallstack&localTrackOrderByPid=7225-0~&thread=0&v=4) - few clicks in a row.

**Comment by gorhill** [comment]
I don't understand "Then only this case". You mean "then only this case" should be fixed?

**Comment by gwarser** [comment]
> You mean "then only this case" should be fixed?

Yes. If nothing can be done to improve this querySelector(All) call, then it should be called as rarely as possible.

**Comment by gorhill** [comment]
How does profiler results compare with before?

**Comment by gwarser** [comment]
Hard to say. On page from OP it's ~2x faster. No, or small load on subsequent clicks, so this part is good. On the other hand on youtube.com or new.reddit.com first call looks very bad in profile. However CPU load seems to be more evenly distributed between CPU cores, so it may be less jerky.

**Comment by gorhill** [comment]
> seems to be more evenly distributed between CPU cores

uBO's content scripts run in the same process as the page's script.

I am not very happy with those results. I will re-open. There does not seem to be solution -- removing the badge count for the cosmetic filtering icon is really what would get rid of the issue; other blockers do not have this feature and thus are not suffering from it.

**Comment by gorhill** [comment]
At this link, <https://www.reddit.com/r/uBlockOrigin/>, default settings in uBO, I get ~40 ms in Chromium stable and ~50 ms in Firefox stable. I use integrated performance profiler for Chromium and Firefox Profiler for Firefox.

**Comment by gwarser** [comment]
> > seems to be more evenly distributed between CPU cores
> 
> uBO's content scripts run in the same process as the page's script.

What I meant, heavy calculation in one large loop/function call can block other tasks - can prevents switching to them. If load is distributed between many calls to "small" function, tasks can be switched more often, which _may_ have good effect on browser responsiveness. 

> I am not very happy with those results. I will re-open. There does not seem to be solution -- removing the badge count for the cosmetic filtering icon is really what would get rid of the issue; other blockers do not have this feature and thus are not suffering from it.

:+1: 


> At this link, [reddit.com/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/), default settings in uBO, I get ~40 ms in Chromium stable and ~50 ms in Firefox stable. I used integrated performance profiler for Chromium and Firefox Profiler for Firefox.

I tried clean profile with https://old.reddit.com/r/uBlockOrigin/ which loads only 25 results by default, and in b5 `getFilteredElementCount` takes 82ms, in b7 `surveyResults.hiddenElementCount` takes 68 ms.
In new.reddit 74ms in b5 and 84ms in b7. But with old code I clearly see spikes in cpu usage, where in new they hide in noise.

***

I need to check why on my daily profile it requires 1-1.5s for this.

Heh ??? It looks like moderation privilege causes uBO subreddit to be 2 to 5 times slower :D Logged as gwarser on https://old.reddit.com/r/uBlockOrigin/ -> 1400ms, logged in my second account -> 300ms, gwarser on https://old.reddit.com/r/firefox/ -> ~500ms, second account -> ~550ms

**Comment by gorhill** [comment]
> If load is distributed between many calls to "small" function

Sure but that is not really what I implemented -- collating all the hidden elements still requires one JS pass.

In any case, the latest changes are:
- For hidden elements: update the badge value on-demand, when hovering over it.
- For inline script tags or other sort of inline JS code: search for these if and only if no other script tags has been found (not in dev build yet).
- For either, the result is cached until the DOM is mutated.

**Comment by uBlock-user** [comment]
On this page and the ones where there are lots of responses on this issue tracker, it takes upto a second or two for the . . . to disappear. Is this normal ? I do have all annoyances lists activated.

**Comment by gwarser** [comment]
> On this page...

Ouch! https://profiler.firefox.com/public/beb9f7cc737e8e2355f123fc7cc6b6b49f2de7d8/calltree/?globalTrackOrder=0&implementation=js&localTrackOrderByPid=1582-0~&thread=0&v=4

**Comment by gorhill** [comment]
> I do have all annoyances lists activated.

That can make a big difference, so far I tested with default lists.

**Comment by gorhill** [comment]
Please, I need to know what URL visited and what filter lists are selected with profiling results, I am unable to guess. I tried this page along with all Annoyances lists enabled and I got 210 ms on Firefox 70.0.1:

https://profiler.firefox.com/public/d51106894cd056d26b97bf13260c2143270d42a7/calltree/?globalTrackOrder=0-1-2&implementation=js&thread=2&v=4

**Comment by uBlock-user** [comment]
Built-In -- all except Unbreak, Ads -- 4/4, Privacy 3/3, Annoyances -- 6/6, 

Custom -- 

```
https://easylist-downloads.adblockplus.org/adwarefilters.txt
https://raw.githubusercontent.com/metaphoricgiraffe/tracking-filters/master/trackingfilters.txt
https://raw.githubusercontent.com/uBlock-user/uBO-Personal-Filters/master/uPF.txt
https://raw.githubusercontent.com/yourduskquibbles/webannoyances/master/filters/generic_filters.txt
https://raw.githubusercontent.com/yourduskquibbles/webannoyances/master/filters/modal_filters.txt
https://raw.githubusercontent.com/yourduskquibbles/webannoyances/master/filters/newsletter_filters.txt
https://raw.githubusercontent.com/yourduskquibbles/webannoyances/master/filters/social_filters.txt 
```

URLs - 

```
https://github.com/uBlockOrigin/uBlock-issues/issues/756
https://facebook.com (probably will need an account)
https://twitter.com/Jack
https://github.com/uBlockOrigin/uBlock-issues/issues/615
```

> profiling results

I didn't profile, jus my observation if you try element count on those pages with aformentioned lists enabled.

**Comment by gorhill** [comment]
> For Twitter keep scrolling on this profile until you reach the very first tweet, will take a long while to reach.

On this one I can reproduce a 4s delay, without needing to go to the very first tweet -- default lists + all Annoyances lists selected + AdGuard Base list.

So the conclusion is simple, uBO needs to bail out after a set time to avoid this, meaning it may happen the badge can't be evaluated due to performance considerations.

**Comment by uBlock-user** [comment]
Can you not reproduce on this page on github ?

**Comment by gorhill** [comment]
> Can you not reproduce on this page on github ?

I [commented above](https://github.com/uBlockOrigin/uBlock-issues/issues/756#issuecomment-549038989):

> I tried this page here along with all Annoyances lists enabled and I got 210 ms on Firefox 70.0.1

**Comment by gwarser** [comment]
@gorhill when you profiled this, were you logged in into your account? I tried in clean Nightly profile, default uBO b9 settings and it's similar case like for Reddit I mentioned above - for `surveyResults.hiddenElementCount<` 184 ms when logged out, but 2134ms when logged in my account.

**Comment by gorhill** [comment]
I understood "on this page" to mean the current page I am typing on.

**Comment by gwarser** [comment]
Yes, this issue page: `https://github.com/uBlockOrigin/uBlock-issues/issues/756`

**Comment by gorhill** [comment]
I could reproduce 3.5s+ when logged in. I logged out, opened the page in a new tab and back to ~270 ms. Odd, I don't know what to make of this. The scriptlet for some reasons spend a whole lot more time executing `matches` and `closest` when I was logged in.

In any case, now uBO bails out after 100ms trying to count hidden elements.

**Comment by uBlock-user** [comment]
on facebook/twitter now, even without scrolling anything it shows `?`straight, very badly designed.

**Comment by gorhill** [comment]
> very badly designed websites.

I won't defend how these sites are designed, the real problem however is having to find out the number of elements affected by a specific set of CSS rules, there is no native API to accomplish this and so we have to do this using potentially slow JS code -- this is not the fault of the web sites.

**Comment by uBlock-user** [comment]
This is not a isolated incident, performance always drops whenever you start scrolling on your newsfeed on facebook, this just happens to be the one of the issues on their website, but on the browser side of things we do need a native API for this.

**Comment by gorhill** [comment]
There is a logic mistake in my last commit, it does not bail out as it's supposed to, it currently needs to find at least one matching element before it bails out, which means the issue is still there for when there is not a single one matching element on a page.

**Comment by troysjanda** [comment]
Can there be a advance setting to revert to showing how many are blocked.  Every site now just displays "?" and I have an 8 core processor. I do not like this change FYI! The roll over "..." is ok but after visiting almost every bookmark and getting just an ?

**Comment by gorhill** [comment]
Advanced setting is not possible, the scriptlet has to execute in a synchronous manner. 8 core processor does not matter here, the scriptlet execute in the content process, i.e. one core. The issue is probably you have a lot of cosmetic filters.

**Comment by troysjanda** [comment]
> Advanced setting is not possible, the scriptlet has to execute in a synchronous manner. 8 core processor does not matter here, the scriptlet execute in the content process, i.e. one core. The issue is probably you have a lot of cosmetic filters.

I use the standard lists that come with Ubo, no custom lists. Besides the default Ubo lists I use easylist easy privacy, and fanboy annoyance, and fanboy cookie. Every site displays a "?" after rolling over the "...", I guess that Ill be stuck using the  v1.23.1b6.

**Comment by gorhill** [comment]
What the issue here says is that I should have never put in in the first place, other blockers do not have this, so currently the fix here is the least worst option, the worst is removing the badge altogether. I can't do magic.

**Comment by troysjanda** [comment]
ill just use the  v1.23.1b6 version until it no longer works. I might be the minority here but I like to see the count numbers, I  read through this issue and I don't use facebook, twitter and I am not seeing any spikes in cpu usage that are a result of Ubo. I understand that you cant please everyone. Just don't like this particular change and Ill make due with the older version until it no longer works.

**Comment by uBlock-user** [comment]
> here is the least worst option

or you could keep the badge but don't show the numbers but instead show the switch bolded or darkend so to signify that something has been blocked, possible ?

**Comment by troysjanda** [comment]
ill just use the v1.23.1b6 version until it no longer works. As I said I might be in a real minority of those that like to see the numbers. I don't want to be that guy that messes up a change that the majority might like ill just continue to use ill just use the v1.23.1b6.

**Comment by gorhill** [comment]
> something has been blocked

On this page here, to find out whether something has been blocked requires not bailing out, meaning 100% CPU usage for seconds as shown in the above profiling results. uBO needs to bail out, there is no magic way to deal with this performance issue.

**Comment by gorhill** [comment]
> the majority might like

It's not even a matter of like/not like, it's a matter of preventing uBO from hogging the CPU for seconds just for the sake of badge count, and potentially interfering seriously with a site's proper functioning.

**Comment by troysjanda** [comment]
Thank you, Maybe its just me but I am not seeing spikes in cpu. This is a great addons and ill continue to use v1.23.1b6 thanks for your time.

**Comment by gwarser** [comment]
After latest comment here, I started getting insane numbers in profiler, like 15 seconds! 

I think I have found optimization - there is no point counting nodes inside already hidden nodes, so maybe something like this: 

```diff
diff --git a/src/js/scriptlets/dom-survey-elements.js b/src/js/scriptlets/dom-survey-elements.js
index e315b8b1..4d5b43cd 100644
--- a/src/js/scriptlets/dom-survey-elements.js
+++ b/src/js/scriptlets/dom-survey-elements.js
@@ -79,6 +79,7 @@
             for (;;) {
                 const node = nodeIter.nextNode();
                 if ( node === null ) { break; }
+                if ( node.parentNode.offsetParent == null ) { continue; }
                 if ( node.offsetParent === null ) {
                     candidates.add(node);
                 }

```

**Comment by gorhill** [comment]
> Advanced setting is not possible

Took a break from computer then it occurred to me I don't need to pass setting to the scriptlet, I just need to inject a different scriptlet, so I can bring back the previous behavior with an advanced setting.

**Comment by gorhill** [comment]
> there is no point counting nodes inside already hidden nodes

I don't understand your thinking: the badge needs to report the number of hidden nodes as a result of injected cosmetic filters. What's the point of testing nodes which are not hidden? Surely they are not affected by cosmetic filters?

**Comment by gorhill** [comment]
> there is no point counting nodes inside already hidden nodes

Good idea (sorry I didn't read properly before my deleted comment). In such case, it might be better if I use a [TreeWalker](https://developer.mozilla.org/en-US/docs/Web/API/Treewalker) instead of a node iterator, if a node is hidden, I just move to the next sibling instead of the next node.

**Comment by gorhill** [comment]
> Good idea

Well actually I don't think this can work, `offsetParent`  can be `null` even when a node is not necessarily affected by `display: none`, see "Notes" at [documentation](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetParent). So maybe you got better result because `offsetParent` for `document.body` is always `null`, meaning you were essentially testing one single node only, `document.body`.

**Comment by gwarser** [comment]
I tested this on https://old.reddit.com/r/uBlockOrigin/ and returned valid numbers (over 50 blocked).

**Comment by gorhill** [comment]
Interesting, 

URL: `https://www.google.ca/search?hl=en&q=laptop`

When I skip descendants in case of match, I get 2 matches on the badge.

Using 1.23.0, I get ~80 matches

Using 1.23.1b12, I get `?`.

So obviously skipping descendants is better than ending up with `?`, though not exactly what was reported before. To achieve this, there will be an advanced setting to restore the old, potentially slow surveyor code.

**Comment by gorhill** [comment]
> returned valid numbers (over 50 blocked).

There would be a problem if any of the cosmetically filtered elements are children of `body` or of a `position: fixed` element -- so we really need to skip descendants only in case of match.

**Comment by gwarser** [comment]
"all" code should be set as default an other code removed. With on-demand approach there is no longer issue with performance every time popup is opened (based on OP description - solved). `querySelectorAll` is still faster than node walking code (when checked with time limit unlocked). Time limit was good, but when requested on demand, can be frustrating to get `?` half the time - I'm willing to wait few seconds to get "exact" numbers when I really want them, but now it's "protected in two steps" by on-demand mouse hover, and then time limit which finally gives `?`.

**Comment by gwarser** [comment]
:+1: 

Thank you.

**Comment by BoneTone** [comment]
I want to mention this user scenario in case some future issue brings up this question of displaying the numbers again.

I don't usually care about the exact number displayed on the badge, but having a number is useful.  It helps to easily spot as well as quantify a certain type of terribly coded website.

There have been a few websites I have encountered that were coded such that if a certain net request was blocked, they would keep retrying, or continuously try slight modifications of the request (essentially incrementing an index in the URL).  On at least one of these sites, the requests were being sent at such a rapid pace it would stress the CPU.

When this atrocious desire to pound a user with trackers was encountered, in just a few minutes time (~2-5), the count would show hundreds of thousands of blocked requests.  On the grammarly site, I saw over 700k once; that's how the badge displays a number this large, with the "k".  I wonder if I let it run a little longer if it would change to an m for millions of blocked requests. 

So it's nice to be able to have that number count all the blocked requests when a site is doing that.  I think the new implementation will still give the same count, as it's not cosmetic filters with children to skip.  But I wanted to make sure this kind of thing would be considered should there be a future change that modifies how the count is created. 

This kind of behavior, by the website, is so offensive in my view that I no longer visit websites that exhibit it, and I will uninstall their software if I have it installed in the first place.

PS - Let me know if there's any additional testing of this change that you would like me to do.  I know it's been fixed for a while now, so just offering if you want or need it.
